library/core/src/any.rs

```diff
@@ -689,7 +690,22 @@ impl TypeId {
     #[stable(feature = "rust1", since = "1.0.0")]
     #[rustc_const_unstable(feature = "const_type_id", issue = "77125")]
     pub const fn of<T: ?Sized + 'static>() -> TypeId {
-        TypeId { t: intrinsics::type_id::<T>() }
+        #[cfg(bootstrap)]
+        let t = intrinsics::type_id::<T>() as u128;
+        #[cfg(not(bootstrap))]
+        let t: u128 = intrinsics::type_id::<T>();
+        TypeId { t }
+    }
+}
+
+#[stable(feature = "rust1", since = "1.0.0")]
+impl hash::Hash for TypeId {
+    #[inline]
+    fn hash<H: ~const hash::Hasher>(&self, state: &mut H) {
+        // Hack: hash as `(u64, u64)` to avoid breaking some libraries used in
+        // the compiler. This should be considered more carefully before landing.
+        (self.t as u64).hash(state);
+        ((self.t >> 64) as u64).hash(state);
```

Comment:
This might work, but IIRC in #95845 (or some other PR), I made it hash just 64 bits.
One thing to note is that this being *a hash*, more bits isn't more better, and you can avoid any perf complaints by only hashing 64 bits just like before.

IOW, it doesn't have to be *perfect*, as `Eq` will still compare the full width, you just need to feed at least 64 bits to the hasher ([`Hasher::finish`](https://doc.rust-lang.org/std/hash/trait.Hasher.html#tymethod.finish) returns `u64` so I suspect that more bits than that is overkill).

Reply:
> This might work, but IIRC in https://github.com/rust-lang/rust/pull/95845 (or some other PR), I made it hash just 64 bits.

Hm, that's reasonable. I feel like it's better to hash the full value, but I don't feel that strongly about it.

Reply:
> [`Hasher::finish`](https://doc.rust-lang.org/std/hash/trait.Hasher.html#tymethod.finish) returns `u64` so I suspect that more bits than that is overkill.

This is a really compelling point, honestly.

Reply:
Compelling enough that I changed it. I still think the impl should get t-libs sign-off though, but not worth doing that prior to the MCP being accepted.